File: backtest/engine.py
# -*- coding: utf-8 -*-
"""
回测引擎：主回测循环。

相关模块：
  - position_handler: 单根 K 线持仓处理
  - signal_scanner: 入场信号扫描
  - trade_helper: 交易记录创建辅助
  - cooldown: 冷却期管理
  - loss_tracker: 连续亏损追踪
"""

from __future__ import annotations

import logging

import pandas as pd
from tqdm import tqdm

from .config import (
    INITIAL_CAPITAL,
    MAX_POSITIONS,
    MIN_OPEN_CASH,
    POSITION_RATIO,
    STOP_LOSS_FACTOR,
    COOLDOWN_HOURS,
)
from .data import build_global_timeline
from .indicators import calc_recent_high, calc_recent_low
from .models import Position, TradeRecord
from .position_handler import process_candle, process_entry_candle
from .signal_scanner import scan_signals
from .cooldown import set_cooldown
from .loss_tracker import LossTracker
from .trade_helper import create_trade_record

logger = logging.getLogger(__name__)


def _execute_pending_buys(
    pending_buys: list[tuple[str, str, float]],
    ts: str,
    data: dict[str, pd.DataFrame],
    ts_to_idx: dict[str, dict[str, int]],
    cash: float,
    portfolio_log: list[tuple[str, float]],
    positions: list[Position],
) -> tuple[list[tuple[str, str, float]], float]:
    """
    执行上一时间步挂单的买入。

    返回：(new_pending_buys, updated_cash)
    """
    new_pending: list[tuple[str, str, float]] = []

    for sym, sig_ts, rr_ratio in pending_buys:
        df = data.get(sym)
        if df is None:
            continue
        idx_map = ts_to_idx.get(sym, {})
        cur_idx = idx_map.get(ts)
        if cur_idx is None:
            new_pending.append((sym, sig_ts, rr_ratio))
            continue

        open_price = float(df.at[cur_idx, "open"])
        last_nav = portfolio_log[-1][1] if portfolio_log else INITIAL_CAPITAL
        position_size = last_nav * POSITION_RATIO
        alloc = min(position_size, cash)
        if alloc < MIN_OPEN_CASH or alloc <= 0:
            continue

        shares = alloc / open_price
        rec_low, _ = calc_recent_low(df, cur_idx)
        stop_p = rec_low * STOP_LOSS_FACTOR

        rec_high, rec_high_time = calc_recent_high(df, cur_idx)
        rec_low, rec_low_time = calc_recent_low(df, cur_idx)
        init_stop_loss_pct = (open_price - rec_low) / open_price * 100 if open_price > 0 else 0.0

        entry_rsn = (
            f"盈亏比 {rr_ratio:.2f}\n"
            f"阶段高点 {rec_high_time} {rec_high:.6g}\n"
            f"阶段低点 {rec_low_time} {rec_low:.6g}\n"
            f"初次止损幅度 {init_stop_loss_pct:.2f}%"
        )

        pos = Position(
            symbol=sym,
            entry_price=open_price,
            entry_time=ts,
            entry_idx=cur_idx,
            shares=shares,
            allocated=alloc,
            stop_price=stop_p,
            recent_high=rec_high,
            max_close=open_price,
            stop_reason="阶段低点止损",
            entry_rr_ratio=rr_ratio,
            recent_high_time=rec_high_time,
            recent_low_time=rec_low_time,
            entry_reason=entry_rsn,
        )
        cash -= alloc
        positions.append(pos)
        logger.debug("买入 %s @ %.6f  止损 %.6f  止盈触发 %.6f", sym, open_price, stop_p, rec_high)

    return new_pending, cash


def _process_positions(
    positions: list[Position],
    ts: str,
    data: dict[str, pd.DataFrame],
    ts_to_idx: dict[str, dict[str, int]],
    cash: float,
    all_trades: list[TradeRecord],
    cooldown_until: dict[str, str],
    loss_tracker: LossTracker,
    ts_to_global_idx: dict[str, int],
) -> tuple[list[Position], float]:
    """
    处理所有持仓，返回存活持仓和更新后的现金。
    """
    surviving: list[Position] = []

    for pos in positions:
        df = data.get(pos.symbol)
        if df is None:
            surviving.append(pos)
            continue
        idx_map = ts_to_idx.get(pos.symbol, {})
        cur_idx = idx_map.get(ts)
        if cur_idx is None:
            surviving.append(pos)
            continue

        # 买入当根特殊处理
        if ts == pos.entry_time:
            cash, trade_recs, exited = process_entry_candle(
                pos, df, cur_idx, ts, cash, cooldown_until
            )
            all_trades.extend(trade_recs)
            if exited:
                # 处理连续亏损
                if trade_recs and not trade_recs[-1].is_half:
                    cur_idx_global = ts_to_global_idx.get(ts, 0)
                    loss_tracker.process_trade(trade_recs[-1], cur_idx_global)
            else:
                surviving.append(pos)
            continue

        # 常规K线处理
        action, cash_delta, trade_recs = process_candle(pos, df, cur_idx, cash)
        cash += cash_delta
        all_trades.extend(trade_recs)

        if action == "exit_full":
            logger.debug(
                "平仓 %s  原因=%s  现金回收=%.2f",
                pos.symbol, trade_recs[-1].exit_reason if trade_recs else "?", cash_delta
            )
            set_cooldown(cooldown_until, pos.symbol, ts)

            # 连续亏损冷却检测
            if trade_recs and not trade_recs[-1].is_half:
                cur_idx_global = ts_to_global_idx.get(ts, 0)
                loss_tracker.process_trade(trade_recs[-1], cur_idx_global)
        else:
            surviving.append(pos)

    return surviving, cash


def _calculate_portfolio_value(
    positions: list[Position],
    ts: str,
    data: dict[str, pd.DataFrame],
    ts_to_idx: dict[str, dict[str, int]],
    cash: float,
) -> tuple[float, list[dict]]:
    """
    计算当前持仓市值和快照。

    返回：(portfolio_value, position_snapshots)
    """
    holding_value = 0.0
    snapshot: list[dict] = []

    for pos in positions:
        df = data.get(pos.symbol)
        if df is None:
            continue
        idx_map = ts_to_idx.get(pos.symbol, {})
        cur_idx = idx_map.get(ts)
        if cur_idx is None:
            continue
        close_price = float(df.at[cur_idx, "close"])
        holding_value += pos.shares * close_price
        pnl_pct = (close_price - pos.entry_price) / pos.entry_price * 100 if pos.entry_price else 0.0
        snapshot.append({
            "symbol": pos.symbol,
            "entry_time": pos.entry_time,
            "hold_h": pos.candle_count,
            "pnl_pct": round(pnl_pct, 2),
        })

    return cash + holding_value, snapshot


def _force_close_positions(
    positions: list[Position],
    timestamps: list[str],
    data: dict[str, pd.DataFrame],
    ts_to_idx: dict[str, dict[str, int]],
    cash: float,
    all_trades: list[TradeRecord],
    loss_tracker: LossTracker,
    ts_to_global_idx: dict[str, int],
) -> float:
    """
    回测结束时强制平仓所有剩余持仓。

    返回：更新后的现金余额
    """
    if not positions or not timestamps:
        return cash

    last_ts = timestamps[-1]

    for pos in positions:
        df = data.get(pos.symbol)
        if df is None:
            continue
        idx_map = ts_to_idx.get(pos.symbol, {})
        cur_idx = idx_map.get(last_ts)
        if cur_idx is None:
            available = [(t, i) for t, i in idx_map.items() if t <= last_ts]
            if not available:
                continue
            cur_idx = max(available, key=lambda x: x[0])[1]

        close_price = float(df.at[cur_idx, "close"])
        proceeds = pos.shares * close_price
        cost_basis = pos.shares * pos.entry_price
        pnl = proceeds - cost_basis
        cash += proceeds

        trade_record = create_trade_record(
            pos, last_ts, close_price, pos.shares, pnl,
            "回测结束", pos.candle_count, is_half=False,
        )
        all_trades.append(trade_record)
        logger.debug(
            "回测结束强制平仓 %s  entry=%s  close=%.6f  pnl=%.2f",
            pos.symbol, pos.entry_time, close_price, pnl
        )

        # 连续亏损冷却检测
        loss_tracker.process_trade(trade_record, ts_to_global_idx.get(last_ts, 0))

    return cash


def run_backtest(
    data: dict[str, pd.DataFrame],
    backtest_start: dict[str, int],
) -> tuple[list[TradeRecord], list[tuple[str, float]], list[list[dict]]]:
    """
    执行完整回测。

    返回：
      all_trades    : 所有已完结的交易记录
      portfolio_log : [(open_time, portfolio_value), ...]
      pos_snapshots : 与 portfolio_log 等长，每项为当根收盘后持仓快照列表
    """
    # 初始化索引映射
    ts_to_idx: dict[str, dict[str, int]] = {}
    for symbol, df in data.items():
        ts_to_idx[symbol] = dict(zip(df["open_time"], df.index))

    timestamps = build_global_timeline(data, backtest_start)
    ts_to_global_idx: dict[str, int] = {ts: i for i, ts in enumerate(timestamps)}

    # 初始化状态
    cash = INITIAL_CAPITAL
    positions: list[Position] = []
    pending_buys: list[tuple[str, str, float]] = []
    all_trades: list[TradeRecord] = []
    portfolio_log: list[tuple[str, float]] = []
    pos_snapshots: list[list[dict]] = []
    cooldown_until: dict[str, str] = {}
    loss_tracker = LossTracker()

    for ts in tqdm(timestamps, desc="回测进度", unit="bar", dynamic_ncols=True):
        # ── 1. 执行上一时间步挂单的买入 ──
        pending_buys, cash = _execute_pending_buys(
            pending_buys, ts, data, ts_to_idx, cash, portfolio_log, positions
        )

        # ── 2. 处理每个持仓 ──
        positions, cash = _process_positions(
            positions, ts, data, ts_to_idx, cash, all_trades,
            cooldown_until, loss_tracker, ts_to_global_idx
        )

        # ── 3. 计算当前持仓市值，同步记录持仓快照 ──
        portfolio_val, snapshot = _calculate_portfolio_value(
            positions, ts, data, ts_to_idx, cash
        )
        portfolio_log.append((ts, portfolio_val))
        pos_snapshots.append(snapshot)

        # ── 4. 判断是否允许开新仓，再按需扫描入场信号 ──
        n_pos = len(positions)
        all_half = (n_pos == MAX_POSITIONS and all(p.half_sold for p in positions))
        allow_new = (n_pos < MAX_POSITIONS) or all_half

        # 检查全局冷却期
        cur_global_idx = ts_to_global_idx.get(ts, 0)
        if loss_tracker.is_in_cooldown(cur_global_idx):
            remaining = loss_tracker.get_remaining_cooldown(cur_global_idx)
            logger.debug("全局冷却中，剩余%d个周期，跳过入场扫描", remaining)

        if allow_new and cash >= MIN_OPEN_CASH and not loss_tracker.is_in_cooldown(cur_global_idx):
            slots_to_fill = (MAX_POSITIONS + 1 - n_pos) if all_half else (MAX_POSITIONS - n_pos)

            if slots_to_fill > 0:
                held_symbols = {p.symbol for p in positions}
                held_symbols |= {sym for sym, _ in pending_buys}
                candidates = scan_signals(data, ts, ts_to_idx, held_symbols, cooldown_until)
                if candidates:
                    sym, rr = candidates[0]
                    pending_buys.append((sym, ts, rr))

    # ── 5. 回测结束：强制平仓所有剩余持仓 ──
    cash = _force_close_positions(
        positions, timestamps, data, ts_to_idx, cash,
        all_trades, loss_tracker, ts_to_global_idx
    )

    return all_trades, portfolio_log, pos_snapshots

File: apps/server/src/backtest/engine/engine.ts
/**
 * 回测引擎主循环 — 精确翻译自 backtest/engine.py
 * 重构点：
 *   1. 冷却机制从 per-symbol setCooldown + LossTracker 改为账户级 CooldownState
 *   2. 主循环新增 candleLog 逐根事件收集
 *   3. 新增凯利公式动态仓位与模拟期逻辑
 */

import { precomputeAllKdj, precomputeBrickChartAll } from './bt-indicators';
import { initCooldown, isInCooldown } from './cooldown';
import { createPosition, type TradeRecord } from './models';
import { createTradeRecord } from './trade-helper';
import { yieldToEventLoop } from './steps/engine.async';
import type { BacktestResult } from './steps/engine.types';
import { forceClosePositions } from './steps/engine.force-close';
import { executePendingBuys, type KellyContext } from './steps/engine.pending-execution';
import { calculateOpenEquity, calculatePortfolioValue } from './steps/engine.portfolio-marks';
import { processPositions } from './steps/engine.position-processing';
import { buildGlobalTimeline } from './steps/engine.timeline';
import type {
  BacktestConfig,
  CandleEntryEvent,
  CandleExitEvent,
  CandleLogEntry,
  KlineBarRow,
  Position,
} from './models';
import { scanSignals } from './signal-scanner';

export type { BacktestResult } from './steps/engine.types';
export { buildGlobalTimeline } from './steps/engine.timeline';

function updateKellyStats(
  simTrades: TradeRecord[],
  allTrades: TradeRecord[],
  windowSize: number,
): { p: number; b: number; cumP: number; cumB: number } {
  const allCompleted = [...simTrades, ...allTrades].filter((t) => !t.isHalf);
  const windowSamples = allCompleted.slice(-windowSize);

  function calcStats(samples: TradeRecord[]) {
    const wins = samples.filter((t) => t.pnl > 0);
    const losses = samples.filter((t) => t.pnl <= 0);
    const p = samples.length ? wins.length / samples.length : 0;

    let b = 0;
    if (losses.length === 0) {
      if (wins.length > 0) b = 999;
    } else if (wins.length === 0) {
      b = 0;
    } else {
      const avgWin = wins.reduce((s, t) => s + (t.overallReturnPct ?? 0), 0) / wins.length;
      const avgLoss = Math.abs(losses.reduce((s, t) => s + (t.overallReturnPct ?? 0), 0) / losses.length);
      b = avgLoss > 0 ? avgWin / avgLoss : 0;
    }
    return { p, b };
  }

  const cum = calcStats(allCompleted);
  const win = calcStats(windowSamples);
  return { p: win.p, b: win.b, cumP: cum.p, cumB: cum.b };
}

function tradeRecsToExitEvents(tradeRecs: TradeRecord[]): CandleExitEvent[] {
  return tradeRecs.map((rec) => ({
    symbol: rec.symbol,
    price: rec.exitPrice,
    shares: rec.shares,
    amount: rec.shares * rec.exitPrice,
    pnl: rec.pnl,
    reason: rec.exitReason,
    isHalf: rec.isHalf,
    isSimulation: rec.isSimulation ?? false,
    tradePhase: rec.tradePhase ?? 'live',
    overallReturnPct: rec.overallReturnPct,
    cumulativeWinRate: rec.cumulativeWinRate,
    cumulativeOdds: rec.cumulativeOdds,
    windowWinRate: rec.windowWinRate,
    windowOdds: rec.windowOdds,
  }));
}

function forceCloseSimPositions(
  simPositions: Position[],
  ts: string,
  data: Map<string, KlineBarRow[]>,
  tsToIdx: Map<string, Map<string, number>>,
  simTrades: TradeRecord[],
  tradePhase: 'simulation' | 'probe',
): TradeRecord[] {
  const reason = tradePhase === 'probe' ? '探针模式结束强平' : '模拟期结束强平';
  const tradeRecs: TradeRecord[] = [];
  for (const pos of simPositions) {
    const df = data.get(pos.symbol);
    if (!df) continue;
    const idxMap = tsToIdx.get(pos.symbol);
    if (!idxMap) continue;
    const curIdx = idxMap.get(ts);
    if (curIdx === undefined) continue;

    const closePrice = df[curIdx].close;
    const proceeds = pos.shares * closePrice;
    const pnl = proceeds - pos.shares * pos.entryPrice;
    const holdCandles = Math.max(1, curIdx - pos.entryIdx + 1);
    const rec = createTradeRecord(pos, ts, closePrice, pos.shares, pnl, reason, holdCandles, false);
    rec.isSimulation = true;
    rec.tradePhase = tradePhase;

    simTrades.push(rec);
    tradeRecs.push(rec);
  }
  return tradeRecs;
}

export async function runBacktest(
  data: Map<string, KlineBarRow[]>,
  backtestStart: Map<string, number>,
  config: BacktestConfig,
  progressCb?: (current: number, total: number, pct: number, currentTs: string) => void,
): Promise<BacktestResult> {
  // 构建 ts → row-index 映射
  const tsToIdx = new Map<string, Map<string, number>>();
  for (const [symbol, df] of data) {
    const m = new Map<string, number>();
    df.forEach((row, i) => m.set(String(row.open_time), i));
    tsToIdx.set(symbol, m);
  }

  // 自定义 KDJ 周期时预计算，避免主循环内重复 O(n²) 计算
  const precomputedKdj =
    config.kdjN !== 9 || config.kdjM1 !== 3 || config.kdjM2 !== 3
      ? precomputeAllKdj(data, config.kdjN, config.kdjM1, config.kdjM2)
      : undefined;

  const brickMap = config.brickXgEnabled ? precomputeBrickChartAll(data) : undefined;

  const timestamps = buildGlobalTimeline(data, backtestStart);

  // 初始化状态
  let cash = config.initialCapital;
  let positions: Position[] = [];
  let pendingBuys: [string, string, number, number][] = [];
  const allTrades: TradeRecord[] = [];
  const portfolioLog: [string, number][] = [];
  const posSnapshots: Array<Array<{ symbol: string; entryTime: string; holdH: number; pnlPct: number }>> = [];
  const candleLog: CandleLogEntry[] = [];

  // 凯利公式模拟期状态
  let simPositions: Position[] = [];
  let simTrades: TradeRecord[] = [];
  let simPortfolioLog: [string, number][] = [];
  let simCash = config.initialCapital;
  let simPendingBuys: [string, string, number, number][] = [];
  let completedTradeCount = 0;
  let currentWindowWinRate = 0;
  let currentWindowOdds = 0;
  const posMeta = new Map<string, { allocated: number; pnlSum: number }>();
  let wasProbeMode = false;

  // 账户级冷却状态（替换旧的 per-symbol cooldownUntil + LossTracker）
  const cooldownState = initCooldown(config.enableCooldown ? config.baseCooldownCandles : 0);

  const totalBars = timestamps.length;
  const REPORT_EVERY = 100;

  const makeKellyCtx = (): KellyContext | undefined => {
    if (!config.enableKellySizing) return undefined;
    return { completedTradeCount, currentWindowWinRate, currentWindowOdds };
  };

  const handleNewTrades = (newTrades: TradeRecord[], phase: 'simulation' | 'probe' | 'live') => {
    const fullExitTrades: TradeRecord[] = [];
    for (const t of newTrades) {
      const key = t.symbol + '|' + t.entryTime;
      const meta = posMeta.get(key);
      if (!meta) continue;
      meta.pnlSum += t.pnl;
      t.isSimulation = phase !== 'live';
      t.tradePhase = phase;
      if (!t.isHalf) {
        t.overallReturnPct = (meta.pnlSum / meta.allocated) * 100;
        posMeta.delete(key);
        fullExitTrades.push(t);
      }
    }

    for (const t of fullExitTrades) {
      completedTradeCount++;
      const stats = updateKellyStats(simTrades, allTrades, config.kellyWindowTrades);
      t.cumulativeWinRate = stats.cumP;
      t.cumulativeOdds = stats.cumB;
      if (completedTradeCount % config.kellyStepTrades === 0) {
        currentWindowWinRate = stats.p;
        currentWindowOdds = stats.b;
      }
      t.windowWinRate = currentWindowWinRate;
      t.windowOdds = currentWindowOdds;
    }
  };

  for (let barIdx = 0; barIdx < timestamps.length; barIdx++) {
    const ts = timestamps[barIdx];
    let entryEvents: CandleEntryEvent[] = [];
    let exitEvents: CandleExitEvent[] = [];

    const isSimPhase = config.enableKellySizing && completedTradeCount < config.kellySimTrades;

    // 计算 kellyRaw（仅在非模拟期且启用凯利时）
    let kellyRaw = 0;
    if (config.enableKellySizing && !isSimPhase) {
      const b = currentWindowOdds;
      const p = currentWindowWinRate;
      if (b > 0 && p > 0) {
        kellyRaw = (b * p - (1 - p)) / b;
      }
    }

    const isProbeMode = config.enableKellySizing && config.enableKellyProbe
      && !isSimPhase
      && kellyRaw <= 0
      && positions.length === 0;

    // 退出 Probe 模式时，强制平掉探针持仓
    if (wasProbeMode && !isProbeMode && simPositions.length > 0) {
      const forcedTrades = forceCloseSimPositions(simPositions, ts, data, tsToIdx, simTrades, 'probe');
      handleNewTrades(forcedTrades, 'probe');
      exitEvents.push(...tradeRecsToExitEvents(forcedTrades));
      simPositions = [];
    }

    // ── 0. 记录 openEquity（在 executePendingBuys 之前，取 open 价格计算） ──
    const openEquity = calculateOpenEquity(positions, ts, data, tsToIdx, cash);

    // ── 1. 执行上一时间步挂单的买入 ──
    if (isSimPhase || isProbeMode) {
      const prevLen = simPositions.length;
      [simPendingBuys, simCash, entryEvents] = executePendingBuys(
        simPendingBuys, ts, data, tsToIdx, simCash, simPortfolioLog, simPositions, config,
        makeKellyCtx(),
        { applyKellySizing: !isProbeMode },
      );
      for (let i = prevLen; i < simPositions.length; i++) {
        const pos = simPositions[i];
        posMeta.set(pos.symbol + '|' + pos.entryTime, { allocated: pos.allocated, pnlSum: 0 });
      }
      const phase = isSimPhase ? 'simulation' : 'probe';
      for (const e of entryEvents) {
        e.isSimulation = true;
        e.tradePhase = phase;
      }
    } else {
      const prevLen = positions.length;
      [pendingBuys, cash, entryEvents] = executePendingBuys(
        pendingBuys, ts, data, tsToIdx, cash, portfolioLog, positions, config, makeKellyCtx(),
      );
      for (let i = prevLen; i < positions.length; i++) {
        const pos = positions[i];
        posMeta.set(pos.symbol + '|' + pos.entryTime, { allocated: pos.allocated, pnlSum: 0 });
      }
      for (const e of entryEvents) {
        e.isSimulation = false;
        e.tradePhase = 'live';
      }
    }

    // ── 2. 处理每个持仓 ──
    if (isSimPhase || isProbeMode) {
      const prevLen = simTrades.length;
      let tradeRecs: TradeRecord[];
      [simPositions, simCash, tradeRecs] = processPositions(
        simPositions, ts, data, tsToIdx, simCash, simTrades,
        cooldownState, barIdx, config, true,
      );
      const phase = isSimPhase ? 'simulation' : 'probe';
      handleNewTrades(simTrades.slice(prevLen), phase);
      exitEvents.push(...tradeRecsToExitEvents(tradeRecs));
    } else {
      const prevLen = allTrades.length;
      let tradeRecs: TradeRecord[];
      [positions, cash, tradeRecs] = processPositions(
        positions, ts, data, tsToIdx, cash, allTrades,
        cooldownState, barIdx, config, false,
      );
      handleNewTrades(allTrades.slice(prevLen), 'live');
      exitEvents.push(...tradeRecsToExitEvents(tradeRecs));
    }

    // 临界点：模拟期结束，强制平掉剩余虚拟持仓
    if (config.enableKellySizing && isSimPhase && completedTradeCount >= config.kellySimTrades) {
      const forcedTrades = forceCloseSimPositions(simPositions, ts, data, tsToIdx, simTrades, 'simulation');
      handleNewTrades(forcedTrades, 'simulation');
      exitEvents.push(...tradeRecsToExitEvents(forcedTrades));
      simPositions = [];
    }

    // ── 3. 计算当前持仓市值，同步记录持仓快照 ──
    if (isSimPhase || isProbeMode) {
      const [simVal] = calculatePortfolioValue(simPositions, ts, data, tsToIdx, simCash);
      simPortfolioLog.push([ts, simVal]);
    }
    const [portfolioVal, snapshot] = calculatePortfolioValue(positions, ts, data, tsToIdx, cash);
    portfolioLog.push([ts, portfolioVal]);
    posSnapshots.push(snapshot);

    // ── 4. 判断是否允许开新仓，再按需扫描入场信号 ──
    const activePositions = isSimPhase ? simPositions : (isProbeMode ? simPositions : positions);
    const nPos = activePositions.length;
    const effectiveMaxPos = config.enableKellySizing ? 1 : config.maxPositions;
    const allowNew = nPos < effectiveMaxPos;

    // 门禁：开启 requireAllPositionsProfitable 时，仅当全部现存持仓满足
    // stopPrice > entryPrice（保本止损已上移至成本之上）方可开新仓；空仓不受限。
    const profitGate =
      !config.requireAllPositionsProfitable ||
      nPos === 0 ||
      activePositions.every((p) => p.stopPrice > p.entryPrice);

    // 账户级冷却门禁（模拟期和探针期不走冷却逻辑）
    const inCooldownNow =
      config.enableCooldown && !isSimPhase && !isProbeMode && isInCooldown(cooldownState, barIdx);

    const activeCash = isSimPhase || isProbeMode ? simCash : cash;
    const activePending = isSimPhase || isProbeMode ? simPendingBuys : pendingBuys;
    if (allowNew && profitGate && activeCash >= config.minOpenCash && !inCooldownNow) {
      const heldSymbols = new Set<string>(activePositions.map((p) => p.symbol));
      for (const [sym] of activePending) heldSymbols.add(sym);
      // scanSignals 不再接收 cooldownUntil 参数
      const candidates = scanSignals(data, ts, tsToIdx, heldSymbols, config, precomputedKdj, brickMap);
      // 有意设计：即使 slotsToFill > 1，也每根只挂 1 单。
      if (candidates.length) {
        const [sym, rr] = candidates[0];
        activePending.push([sym, ts, rr, 0]);
      }
    }

    // ── 5. 推送当根 K 线日志 ──
    // inCooldown 取扫描入场信号时的状态（上方已查询）
    candleLog.push({
      barIdx,
      ts,
      openEquity,
      closeEquity: portfolioVal,
      posCount: positions.length,
      maxPositions: config.enableKellySizing ? 1 : config.maxPositions,
      entries: entryEvents,
      exits: exitEvents,
      openSymbols: positions.map((p) => p.symbol),
      inCooldown: inCooldownNow,
      cooldownDuration: config.enableCooldown ? cooldownState.cooldownDuration : null,
      cooldownRemaining: config.enableCooldown
        ? (inCooldownNow && cooldownState.cooldownUntilBarIdx !== null
            ? cooldownState.cooldownUntilBarIdx - barIdx
            : 0)
        : null,
    });

    wasProbeMode = isProbeMode;

    // 进度回调 + 让出事件循环，确保轮询请求可被处理
    if (barIdx % REPORT_EVERY === 0 || barIdx === totalBars - 1) {
      const pct = totalBars ? ((barIdx + 1) / totalBars) * 100 : 100;
      if (progressCb) progressCb(barIdx + 1, totalBars, pct, ts);
      await yieldToEventLoop();
    }
  }

  // ── 6. 回测结束：强制平仓所有剩余持仓 ──
  // 先处理模拟持仓（若还有剩余）
  if (simPositions.length > 0) {
    const lastTs = timestamps[timestamps.length - 1];
    const forcedTrades = forceCloseSimPositions(simPositions, lastTs, data, tsToIdx, simTrades, wasProbeMode ? 'probe' : 'simulation');
    handleNewTrades(forcedTrades, wasProbeMode ? 'probe' : 'simulation');
    simPositions = [];
  }

  // forceClosePositions 产生的 trade 也走 registerExit，但不 push candleLog（主循环已结束）
  cash = forceClosePositions(
    positions, timestamps, data, tsToIdx, cash, allTrades,
    cooldownState, timestamps.length - 1, config,
  );

  return {
    trades: allTrades,
    simTrades,
    portfolioLog,
    simPortfolioLog,
    posSnapshots,
    candleLog,
  };
}
